miden-tx/src/host/tx_progress.rs

```diff
@@ -0,0 +1,176 @@
+pub use alloc::{
+    string::{String, ToString},
+    vec::Vec,
+};
+
+use miden_objects::notes::NoteId;
+
+// CONSTANTS
+// ================================================================================================
+
+/// Number of cycles needed to create an empty span whithout changing the stack state.
+const SPAN_CREATION_SHIFT: u32 = 2;
+
+// TRANSACTION PROGRESS
+// ================================================================================================
+
+/// Contains the information about the number of cycles for each of the transaction execution
+/// stages.
+#[derive(Default)]
+pub struct TransactionProgress {
+    prologue: CycleInterval,
+    notes_processing: CycleInterval,
+    note_execution: Vec<(Option<NoteId>, CycleInterval)>,
+    tx_script_processing: CycleInterval,
+    epilogue: CycleInterval,
+}
+
+impl TransactionProgress {
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn start_prologue(&mut self, cycle: u32) {
+        self.prologue.set_start(cycle);
+    }
+
+    pub fn end_prologue(&mut self, cycle: u32) {
+        self.prologue.set_end(cycle);
+    }
+
+    pub fn start_notes_processing(&mut self, cycle: u32) {
+        self.notes_processing.set_start(cycle);
+    }
+
+    pub fn end_notes_processing(&mut self, cycle: u32) {
+        self.notes_processing.set_end(cycle);
+    }
+
+    pub fn start_note_execution(&mut self, cycle: u32, note_id: Option<NoteId>) {
+        self.note_execution.push((note_id, CycleInterval::new(cycle)));
+    }
+
+    pub fn end_note_execution(&mut self, cycle: u32) {
+        if let Some((_, interval)) = self.note_execution.last_mut() {
+            interval.set_end(cycle)
+        }
+    }
+
+    pub fn start_tx_script_processing(&mut self, cycle: u32) {
+        self.tx_script_processing.set_start(cycle);
+    }
+
+    pub fn end_tx_script_processing(&mut self, cycle: u32) {
+        self.tx_script_processing.set_end(cycle);
+    }
+
+    pub fn start_epilogue(&mut self, cycle: u32) {
+        self.epilogue.set_start(cycle);
+    }
+
+    pub fn end_epilogue(&mut self, cycle: u32) {
+        self.epilogue.set_end(cycle);
+    }
+
+    // DATA PRINT
+    // --------------------------------------------------------------------------------------------
+
+    pub fn to_json_string(&self) -> String {
```

Comment:
I would probably move this code to the benchmarks crate (we could make all fields in this struct `pub` to make them easily accessible).

Also, since we are using `serde_json` there anyway, we could just build a wrapper struct and serialize it to JSON (should be less code than this).

Reply:
I'm not sure that in this case it will be better to build a wrapper, because:
- We will need to create wrappers for `TransactionProgress`, `CycleInterval` and `NoteId`, overwriting `Serialize::serialize` for each of them. 
- The resulting JSON should contain not just fields of `TransactionProgress`, but calculated length of corresponding intervals, so just a serialized struct is not what we want to get.

So I think that in that case having a custom function to assemble a JSON is most simple, easy and brief solution, although being a bit messy.

Reply:
I think the struct could look something like:
```Rust
pub struct TransactionBenchmark {
    prologue: u32,
    notes_processing: u32,
    note_execution: Vec<(String, u32)>,
    tx_script_processing: u32,
    epilogue: u32,
}
```
And then we can implement `TryFrom<TransactionProgress> for TransactionBenchmark`. This conversion should be pretty straight-forward.